README.md
# Introduction

I love computer science and new technologies to play with. Besides that I like to keep things simple and pay attention
to aspects that really mather. But I also like sharing my world and thoughts with people not involved in computer science.
In fact technology is not dehumanizing. It's what makes us human.


### Why? Don't you have better things to do?

The whole project was basically a wish I had during some **network analysis**:
I wanted to **visualize** my results in some fancier way. There were a lot of dependencies between several hosts
and I'm not good at reading raw packet data. And since **graphs** are an excellent data structure to represent
(computer) networks I've decided to begin my journey :smirk:

During some [D3](http://d3js.org/) experiments  I wanted to capture data from the command line and send it to my D3 application.
Of course I could have read the data from some CSV, JSON, whatever file but that wasn't "real-time" enough. So I had a

* data source
* D3 code applying some magic to the data

The *data source* in that case were some grep lines transformed to JSON data. In fact I was missing the
linking part between mentioned components.


### D3? Never heard of!

**D3** stands for Data-Driven Documents. Its basically a **JavaScript** library aimed to manipulate "documents based on data".
A neat side-effect of using JavaScript: You can use a **browser** to visualize your data. No need for extra GUI, clients whatever.
There are douzens of useful examples how to use D3, especially force layouts which are ment to implement graphs.

# Meet netgrafio!

**netgrafio** provides more or less tools and libraries to visualize your data regardless of its type.
I tried to make the libraries OOP-friendly but as you may know: JS and OOP is really a pain.
In fact I had to code a lot of JS (for the first time in my life as a coder:D).
Its not perfect and it might be buggy. But it works for me :v:

![netgrafio-first-page.png](http://dl.dornea.nu/img/netgrafio/netgrafio-first-page.png)

## Documentation

Make sure you'll have a look at the official documentation available at [http://netgrafio.readthedocs.org/](http://netgrafio.readthedocs.org/).


### Screenshots

![netgrafio-module-analysis.png](http://dl.dornea.nu/img/netgrafio/netgrafio-module-analysis.png)
The network analysis module.

![netgrafio-module-nmap.png](http://dl.dornea.nu/img/netgrafio/netgrafio-module-nmap-results.png)
Visualize nmap results.

### Videos

[![netgrafio - traceroute module](http://img.youtube.com/vi/TpgvQqtE0vU/0.jpg)](https://www.youtube.com/watch?v=TpgvQqtE0vU)

[![netgrafio - network analysis module](http://img.youtube.com/vi/GgKKr3pH1go/0.jpg)](https://www.youtube.com/watch?v=GgKKr3pH1go)

[![netgrafio - nmap results module](http://img.youtube.com/vi/5XhawZkFd-8/0.jpg)](https://www.youtube.com/watch?v=5XhawZkFd-8)


### Basic idea

*netgrafio* listens for some data on a **TCP** socket and passes it through some **WebSocket** to all connected web clients.
The web clients have some JS code which will receive the incoming data and apply some magic to it.

> At the moment only JSON data is supported.

In order to transfer data from one socket to the another a (deadlock free) queue is being used (**producer and consumer pattern**).


### Run me

> All the relevant steps are shown in this [showterm](http://showterm.io/819e4458f5c3b43c84e8d).


#### System requirements

In order to run netgrafio several requirements on your system have to be met. Basically you'll need:

* Python 3.x
* [virtualenv](http://www.virtualenv.org)
* [PIP](http://www.pip-installer.org/)


#### Clone project

Make sure you have installed these packages on your system. Afterwards you can clone this project:

```bash
$ git clone https://github.com/nullsecuritynet/netgrafio
$ cd netgrafio
```

#### Setup virtualenv

Now you'll need to setup a isolated python environment using *virtualenv*:

```bash
$ virtualenv env
Using base prefix '/usr'
New python executable in env/bin/python3
Also creating executable in env/bin/python
Installing setuptools, pip...done.
```

Make sure to activate the virtual environment:

```bash
$ source env/bin/activate
```


#### Install additional packages

Having set the virtualenv environment let's install some missing packages:

```bash
$ pip install -r env/requirements.pip
```


#### Fire up netgrafio

Now you're ready to start netgrafion and have some fun.

These are the basic parameters:

```bash
python netgrafio.py -h
usage: netgrafio.py [-h] [--tcp-port TCP_PORT] [--ws-port WS_PORT]
                    [--web-port WEB_PORT] [--host HOST]

netgrafio - visualize your network

optional arguments:
  -h, --help           show this help message and exit
  --tcp-port TCP_PORT  Specify TCP port to listen for JSON packets (default:
                       8081)
  --ws-port WS_PORT    Specify WebSocket port to send JSON data to (default:
                       8080)
  --web-port WEB_PORT  Specify web port to server web application (default:
                       5000)
  --host HOST          Specify host to bind socket on (default: 127.0.0.1)

```

If you start netgrafio without any arguments, then you'll have a

* *TCP-Socket* listening on port 8081
* *WebSocket* listening on port 8080
* *Web-Application* available at http://localhost:5000


```bash
$ python netgrafio.py
2014-04-24 16:18:12,984 - INFO - [WebSocketServer] - Starting WebSocket server on port 8080
2014-04-24 16:18:12,984 - INFO - [WebSocketServer] - Start collector server
2014-04-24 16:18:12,985 - INFO - [WebSocketServer] - Waiting for incoming data ...
2014-04-24 16:18:12,989 - INFO - [WebServer] - Listening on 5000
2014-04-24 16:18:12,989 - INFO - [TCPServer] - Listening on 8081
```

Now open your browser and navigate to [http://localhost:5000](http://localhost:5000)



CHANGELOG.md
# Changelog

## v0.1

* First version :)

netgrafio.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: victor
# @Date:   2014-04-22
# @Last Modified by:   victor
# @Last Modified time: 2014-04-23
# @Copyright:
#
#    This file is part of the netgrafio project.
#
#
#    Copyright (c) 2014 Victor Dorneanu <info AAET dornea DOT nu >
#
#    Permission is hereby granted, free of charge, to any person obtaining a copy
#    of this software and associated documentation files (the "Software"), to deal
#    in the Software without restriction, including without limitation the rights
#    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#    copies of the Software, and to permit persons to whom the Software is
#    furnished to do so, subject to the following conditions:
#
#    The above copyright notice and this permission notice shall be included in all
#    copies or substantial portions of the Software.
#
#    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#    SOFTWARE.
#
#    The MIT License (MIT)

import sys
import argparse
import logging
from queue import Queue

# Local packages
from lib.TCPServer import JSONServer
from lib.WebSocketServer import WebSocketServer
from lib.WebServer import WebServer


def parse_args(params):
    """ Parse cmd line arguments """
    parser = argparse.ArgumentParser(
        description="netgrafio - visualize your network",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    # Set parameters
    parser.add_argument("--tcp-port", action="store", type=int,
                        help="Specify TCP port to listen for JSON packets", default=8081)
    parser.add_argument("--ws-port", action="store", type=int,
                        help="Specify WebSocket port to send JSON data to", default=8080)
    parser.add_argument("--web-port", action="store", type=str,
                        help="Specify web port to server web application", default=5000)
    parser.add_argument("--host", action="store", default="127.0.0.1",
                        help="Specify host to bind socket on")
    args = parser.parse_args(params)

    return args


def main(params):
    # Global logging settings
    logging.basicConfig(level=logging.DEBUG,
                        format="%(asctime)s - %(levelname)s - [%(name)s] - %(message)s")

    # Init in queue (producer and consumer pattern)
    in_queue = Queue()

    # Set default host
    host = params.host

    # Start WebSocket server
    websocket_server = WebSocketServer(host, int(params.ws_port), in_queue)
    websocket_server.start()

    # Start Web Server
    web_server = WebServer(host, int(params.web_port))
    web_server.start()

    # Start JSON server
    tcp_server = JSONServer(host, int(params.tcp_port), in_queue)
    tcp_server.start()


if __name__ == "__main__":
    main(parse_args(sys.argv[1:]))

# EOF